File: qeli-android/app/src/main/kotlin/com/qeli/SniCatalog.kt
package com.qeli

import android.content.Context
import java.util.Locale
import java.util.TreeSet

/** Built-in + user SNI / TLS front-host catalog (mirrors Qeli.Shared.SniCatalog). */
object SniCatalog {
    @Volatile private var builtin: List<String>? = null

    fun builtin(context: Context): List<String> {
        builtin?.let { return it }
        val loaded = try {
            context.assets.open("sni_hosts.txt").bufferedReader().useLines { lines ->
                lines.mapNotNull { normalize(it) }.distinct().sorted().toList()
            }
        } catch (_: Exception) {
            TransportAuto.SNI_PRESETS
        }
        builtin = loaded
        return loaded
    }

    fun normalize(raw: String?): String? {
        if (raw.isNullOrBlank()) return null
        var h = raw.trim().lowercase(Locale.US)
        val hash = h.indexOf('#')
        if (hash >= 0) h = h.substring(0, hash).trim()
        if (h.isEmpty() || h.startsWith(';')) return null
        if (h.contains("://")) h = h.substringAfter("://")
        h = h.substringBefore('/')
        val colon = h.lastIndexOf(':')
        if (colon > 0 && h.substring(colon + 1).all { it.isDigit() }) {
            h = h.substring(0, colon)
        }
        h = h.trim('.')
        if (h.length !in 3..80 || !h.contains('.')) return null
        return h
    }

    fun merge(context: Context, custom: Collection<String>?): List<String> {
        val set = TreeSet(String.CASE_INSENSITIVE_ORDER)
        set.addAll(builtin(context))
        custom?.forEach { normalize(it)?.let(set::add) }
        return set.toList()
    }
}

File: qeli-shared/QeliShared/SniCatalog.cs
using System.Reflection;

namespace Qeli.Shared;

/// <summary>Built-in + user SNI / TLS front-host catalog.</summary>
public static class SniCatalog
{
    private static IReadOnlyList<string>? _builtin;

    /// <summary>Hosts shipped with the app (community Reality SNI lists + CDN fronts).</summary>
    public static IReadOnlyList<string> Builtin
    {
        get
        {
            if (_builtin != null) return _builtin;
            var asm = typeof(SniCatalog).Assembly;
            var name = asm.GetManifestResourceNames()
                .FirstOrDefault(n => n.EndsWith("sni_hosts.txt", StringComparison.OrdinalIgnoreCase));
            if (name == null)
            {
                _builtin = TransportAuto.SniPresets;
                return _builtin;
            }
            using var stream = asm.GetManifestResourceStream(name)!;
            using var reader = new StreamReader(stream);
            var list = new List<string>();
            while (reader.ReadLine() is { } line)
            {
                var host = Normalize(line);
                if (host != null) list.Add(host);
            }
            _builtin = list.Distinct(StringComparer.OrdinalIgnoreCase).OrderBy(h => h).ToList();
            return _builtin;
        }
    }

    public static string? Normalize(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return null;
        var h = raw.Trim().ToLowerInvariant();
        var hash = h.IndexOf('#');
        if (hash >= 0) h = h[..hash].Trim();
        if (h.Length == 0 || h.StartsWith(';')) return null;
        if (h.Contains("://", StringComparison.Ordinal))
            h = h.Split(new[] { "://" }, 2, StringSplitOptions.None)[1];
        h = h.Split('/')[0];
        var colon = h.LastIndexOf(':');
        if (colon > 0 && h[(colon + 1)..].All(char.IsDigit))
            h = h[..colon];
        h = h.Trim('.');
        if (h.Length is < 3 or > 80) return null;
        if (!h.Contains('.')) return null;
        return h;
    }

    public static IReadOnlyList<string> Merge(IEnumerable<string>? custom)
    {
        var set = new SortedSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var h in Builtin)
            set.Add(h);
        if (custom != null)
        {
            foreach (var c in custom)
            {
                var n = Normalize(c);
                if (n != null) set.Add(n);
            }
        }
        return set.ToList();
    }
}
